### PR TITLE
Fix create entities validation

### LINKED ISSUE #2326: Invalid query is considered valid
RedisGraph 2.8.11

The following invalid query is considered valid:

```
GRAPH.QUERY g "CREATE (n1)-[r:R]->(n2), (n3)-[r:R]->(n2)"
```

Expected error: Variable `r` already declared.

Note that the query creates only one relationship.

### PR DESCRIPTION
**Pull request by alonre24**
Closes #2326, #2306, #2307
Fix bug in `_Validate_CREATE_Entities` (due a typo) -  a "re-declared alias" error was raised for a relation defined in some CREATE clause, instead of being raised if the alias was declared in previous clauses.
Also, perform the validation over previous expressions in the same CREATE clause as well, and over redeclration of nodes in CREATE clause which is not used to create a path.

### CODE REVIEW

File: src/ast/ast_validations.c [hunk 1]
```diff
@@ -906,55 +906,72 @@ static AST_Validation _Validate_MERGE_Clauses(const AST *ast) {
...
-					ErrorCtx_SetError("The bound variable %s' can't be redeclared in a CREATE clause", alias);
-					return AST_INVALID;
-				}
-			}
-
-			// Validate that each relation has exactly one type.
-			uint reltype_count = cypher_ast_rel_pattern_nreltypes(rel);
-			if(reltype_count != 1) {
-				ErrorCtx_SetError("Exactly one relationship type must be specified for CREATE");
+	if(_ValidateInlinedPropertiesOnPath(path) != AST_VALID) return AST_INVALID;
```
Comment (AviAvni) [inline src/ast/ast_validations.c:912]: add empty line after

File: src/ast/ast_validations.c [hunk 2]
```diff
@@ -906,55 +906,72 @@ static AST_Validation _Validate_MERGE_Clauses(const AST *ast) {
...
+	uint nelems = cypher_ast_pattern_path_nelements(path);
+	/* Visit every relationship (every odd offset) on the path to validate its alias and structure.
+	 * TODO There should also be a syntax error for redeclaring nodes, as in:
+	 * MATCH (a) CREATE (a)
+	 * But this is a no-op query, and we don't have the logic to differentiate this from a valid query like
+	 * MATCH (a) CREATE (a)-[:E]->(:B) */
```
Comment (AviAvni) [inline src/ast/ast_validations.c:918]: is this fixed?

File: src/ast/ast_validations.c [hunk 3]
```diff
@@ -906,55 +906,72 @@ static AST_Validation _Validate_MERGE_Clauses(const AST *ast) {
...
+		if(cypher_ast_rel_pattern_get_direction(rel) == CYPHER_REL_BIDIRECTIONAL) {
+			ErrorCtx_SetError("Only directed relationships are supported in CREATE");
+			return AST_INVALID;
 		}
 	}
 
 	return AST_VALID;
 }
 
+static void _AST_Path_GetDefinedIdentifiers(const cypher_astnode_t *path, rax *identifiers) {
```
Comment (AviAvni) [inline src/ast/ast_validations.c:948]: no need to move this function

File: src/ast/ast_validations.c [hunk 4]
```diff
@@ -969,15 +986,27 @@ static AST_Validation _Validate_CREATE_Clauses(const AST *ast) {
 		uint clause_idx = create_clause_indices[i];
 
 		// Collect all entities that are bound before this CREATE clause.
-		for(uint j = start_offset; j < clause_idx; j ++) {
-			const cypher_astnode_t *prev_clause = cypher_ast_query_get_clause(ast->root, i);
+		for (uint j = start_offset; j < clause_idx; j++) {
+			const cypher_astnode_t *prev_clause = cypher_ast_query_get_clause(ast->root, j);
```
Comment (AviAvni) [inline src/ast/ast_validations.c:990]: great catch

File: tests/flow/test_graph_create.py [hunk 5]
```diff
@@ -126,3 +126,20 @@ def test08_create_node_with_2_attr_same_name(self):
...
+    # "variable redeclared" error should return only if the relation alias was already declared
+    def test09_create_use_alias_in_many_clauses(self):
+        query = """CREATE (n1:Node1) CREATE (n2:Node1) CREATE (n1)-[r:Rel1]->(n2)"""
+        result = redis_graph.query(query)
+        self.env.assertEquals(result.nodes_created, 2)
+        self.env.assertEquals(result.relationships_created, 1)
+
+        queries = ["CREATE (n1)-[r:Rel1]->(n2) CREATE (n2)-[r:Rel1]->(n1)",
+                   "CREATE (n1)-[r:Rel1]->(n2), (n2)-[r:Rel1]->(n1)",
+                   "CREATE (n1) CREATE (n1)-[r:Rel1]->(n2), (n3)-[r:Rel1]->(n1)"]
```
Comment (AviAvni) [inline tests/flow/test_graph_create.py:139]: 3rd test is same as first one
make sure all cases covered

**Review by AviAvni — CHANGES_REQUESTED** [review]
thanks, see comments

File: src/ast/ast_validations.c (hunk 2 above)
Comment (alonre24) [reply src/ast/ast_validations.c:918]: No. Do you want me to address this in the scope of this PR?

File: src/ast/ast_validations.c (hunk 3 above)
Comment (alonre24) [reply src/ast/ast_validations.c:948]: I added a forward declaration so we wont have to move it

File: src/ast/ast_validations.c (hunk 2 above)
Comment (AviAvni) [reply src/ast/ast_validations.c:920]: if you think is easy yes please

File: tests/flow/test_graph_create.py [hunk 6]
```diff
@@ -126,3 +126,28 @@ def test08_create_node_with_2_attr_same_name(self):
...
+        # This is a valid query, even though 'n1' is redeclared in the CREATE clause
+        query = """MATCH (n1:Node1) CREATE (n1)-[r:Rel1]->(n2)"""
+        result = redis_graph.query(query)
+        self.env.assertEquals(result.nodes_created, 2)
+        self.env.assertEquals(result.relationships_created, 2)
+
+        queries = ["CREATE (n1)-[r:Rel1]->(n2) CREATE (n2)-[r:Rel1]->(n1)",
+                   "CREATE (n1)-[r:Rel1]->(n2), (n2)-[r:Rel1]->(n1)",
+                   "CREATE (n1)-[r:Rel1]->(n2) CREATE (n2)-[r2:Rel1]->(n3), (n3)-[r:Rel1]->(n2)",
+                   "MATCH (r) CREATE (r)"]
```
Comment (AviAvni) [inline tests/flow/test_graph_create.py:147]: MATCH (r) WITH r CREATE (r) also need to be added and implemented

File: src/ast/ast_validations.c [hunk 7]
```diff
@@ -80,6 +80,7 @@ static bool _ValidateAllShortestPaths
 }
 
 // Forward declaration
+static void _AST_Path_GetDefinedIdentifiers(const cypher_astnode_t *path, rax *identifiers);
 static void _AST_GetDefinedIdentifiers(const cypher_astnode_t *node, rax *identifiers);
```
Comment (AviAvni) [inline src/ast/ast_validations.c:84]: is this 2 functions isn't the same?

**Review by AviAvni — APPROVED** [review]

**Pull request merged**